packages/assets-controllers/src/TokenDetectionController.ts

```diff
@@ -298,27 +428,45 @@ export class TokenDetectionController extends StaticIntervalPollingController<
     networkClientId?: NetworkClientId;
     accountAddress?: string;
   } = {}): Promise<void> {
-    if (
-      this.#disabled ||
-      !this.#isDetectionEnabledForNetwork ||
-      !this.#isDetectionEnabledFromPreferences
-    ) {
+    if (!this.isActive || !this.#isDetectionEnabledForNetwork) {
       return;
     }
-    const { tokens } = this.#getTokensState();
-    const selectedAddress = accountAddress || this.#selectedAddress;
+    const selectedAddress = accountAddress ?? this.#selectedAddress;
     const chainId = this.#getCorrectChainId(networkClientId);
 
-    const tokensAddresses = tokens.map(
-      /* istanbul ignore next*/ (token) => token.address.toLowerCase(),
-    );
+    if (
+      !this.#isDetectionEnabledFromPreferences &&
+      chainId !== ChainId.mainnet
+    ) {
+      return;
+    }
+    const isTokenDetectionInactiveInMainnet =
+      !this.#isDetectionEnabledFromPreferences && chainId === ChainId.mainnet;
     const { tokenList } = this.messagingSystem.call(
       'TokenListController:getState',
     );
+    const tokenListUsed = isTokenDetectionInactiveInMainnet
+      ? STATIC_MAINNET_TOKEN_LIST
+      : tokenList;
+
+    const { tokens, detectedTokens } = this.#getTokensState();
     const tokensToDetect: string[] = [];
-    for (const address of Object.keys(tokenList)) {
-      if (!tokensAddresses.includes(address)) {
-        tokensToDetect.push(address);
+
+    const findCaseInsensitiveMatch = (source: string[], target: string) =>
```

Comment:
What are your thoughts on moving this function outside of the controller?

Reply:
Makes sense! Applied here: [a8f3d05](https://github.com/MetaMask/core/pull/3775/commits/a8f3d0529a80043eb4834a10758a03f8d71e721b)